juju/juju
# Batch Pings() before writing them to the Database

## Description of change

In our existing code, every Ping() that is being driven on every agent causes us to do a Session.Copy() and send a `$inc` to the database. However, its fairly obvious that we could aggregate those pings at some interval. This has some big wins:
1) Because we are setting bits inside integers, we can OR those bit flips in memory and only send a single "$inc" to the database. (So if we have 50 pings for the same slot, we can send a single $inc rather than 50 $inc).
2) This also helps when hitting different fields. We still have to set values for all of those fields, but we can do so for all fields that we see in one pass. (it doesn't change the number of $inc, but it does change the number of times we touch the same document.)
3) The code is written to handle pings across model UUIDs, but we currently create a PingBatcher for each State. This is less of a win because the content is stored in different documents anyway. It would still probably be good, but I'm happy with 1 & 2.
4) Should help tremendously with thundering-herd. The existing Ping() code did a Session.Copy() for every Ping. While many of those will go back into the pool, that still gives a very wide number of concurrent requests possible. (2500 agents => 2500 potential concurrent requests.)
With this, we should significantly reduce contention, because we can handle 2500 concurrent requests as just a mutex + map update, and then 1s later actually flush all of those requests to the database in a single update.
5) The Pruner test "TestDeepStressStaysSane" exists to test our Pruning code plays nicely when lots of pingers are active. That test shows really good (if slightly artificial) results. At 500 pingers (actuall 1k concurrent), the test drops from 2s to 1.5s total time. At 5000 pingers, the test scales to about 18s, and at 10,000 pingers it goes to 37s (essentially linear). With 5000 pingers and the old code, the test never gets past the first loop after 10 minutes. I think that's mostly a mongo socket contention as it doesn't like 5k (10k?) concurrent sessions.
6) I live tested this with a 1s flush speed. I tried to get 1000 'ubuntu' agents running on LXD provider with 8GB RAM, but I kept getting oom-kills of jujud (even 'vim' reading machine-0.log). So I switched and stopped 2 of my containers to end up with 600 running agents. At that point, once it got fully stable (quiet). I saw:
```
                        ns    total    read    write    2017-06-21T15:23:24Z
   presence.presence.pings    211ms     1ms    210ms
            local.oplog.rs     83ms    83ms      0ms
               juju.leases      5ms     2ms      2ms
...
                        ns    total    read    write    2017-06-21T15:23:55Z
   presence.presence.pings    214ms     1ms    213ms
            local.oplog.rs     86ms    86ms      0ms
                 juju.txns      5ms     0ms      5ms
               juju.leases      4ms     2ms      2ms
...
                        ns    total    read    write    2017-06-21T15:24:25Z
   presence.presence.pings    207ms     1ms    206ms
            local.oplog.rs     83ms    83ms      0ms
               juju.leases      5ms     3ms      2ms
...
```

Turned into
```
                        ns    total    read    write    2017-06-21T15:29:22Z
            local.oplog.rs     99ms    99ms      0ms
   presence.presence.pings     12ms     1ms     11ms
               juju.leases      4ms     1ms      2ms
...
                        ns    total     read    write    2017-06-21T15:29:52Z
            local.oplog.rs    105ms    105ms      0ms
   presence.presence.pings     11ms      1ms     10ms
               juju.leases      6ms      3ms      2ms
...
                        ns    total     read    write    2017-06-21T15:29:52Z
            local.oplog.rs    105ms    105ms      0ms
   presence.presence.pings     11ms      1ms     10ms
               juju.leases      6ms      3ms      2ms
```

So that is 'presence.pings' dropping from 200ms / 30s down to 11ms/30s. I also saw times in the first part where pings got huge (, but I can't trust it because we were OOMing.
```
                        ns        total         read        write    2017-06-21T14:58:32Z
   presence.presence.pings    1420615ms         34ms    1420580ms
         juju.applications    1097371ms    1097370ms          0ms
                juju.units     915732ms     915643ms         89ms
             juju.machines     331750ms     331730ms         19ms
```

I decided not to expose a configuration knob. I tried it, and tried several different values. I'm very happy with 1s. We could even go to 10s probably, but 1s means we never wait very log for an agent to show up as 'available', and it makes the load on the database *very* predictable even at 1s frequency.

There are 2 updates to 'update-status-hook-interval' tests because I was cribbing from them and found them to be slightly lacking. Originally included here because I was also changing that area. Still included here cause its worth fixing.

## QA steps

This shouldn't change users interaction with the system. So its mostly that "CI continues to pass all tests". We *should* see scale testing show less database operations being run on the presence.presence.pings collection, and a decrease in things like "mongotop" of that being a high impact collection.

## Documentation changes

Nothing to the user workflow. Only real documentation is performance improvements.

## Bug reference

[lp:1697519](https://bugs.launchpad.net/juju/+bug/1697519)
